src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```diff
@@ -381,63 +432,89 @@ public void Trim(uint tickCount, int id, Utilities.MemoryPressure pressure, int
+                    }
 
-                        while (_count > 0 && trimCount-- > 0)
-                        {
-                            T[]? array = _arrays[--_count];
-                            Debug.Assert(array is not null, "No nulls should have been present in slots < _count.");
-                            _arrays[_count] = null;
+                    if (_timestamp <= tickCount && (tickCount - _timestamp) <= trimTicks)
+                    {
+                        return;
+                    }
+
+                    // We've wrapped the tick count or elapsed enough time since the
+                    // first item went into the stack. Drop the top item so it can
+                    // be collected and make the stack look a little newer.
 
-                            if (log.IsEnabled())
+                    ArrayPoolEventSource log = ArrayPoolEventSource.Log;
+                    int trimCount = StackLowTrimCount;
+                    switch (pressure)
+                    {
+                        case Utilities.MemoryPressure.High:
+                            trimCount = StackHighTrimCount;
+
+                            // When pressure is high, aggressively trim larger arrays.
+                            if (bucketSize > StackLargeBucket)
                             {
-                                log.BufferTrimmed(array.GetHashCode(), array.Length, id);
+                                trimCount++;
                             }
-                        }
+                            if (Unsafe.SizeOf<T>() > StackModerateTypeSize)
+                            {
+                                trimCount++;
+                            }
+                            if (Unsafe.SizeOf<T>() > StackLargeTypeSize)
+                            {
+                                trimCount++;
+                            }
+                            break;
 
-                        if (_count > 0 && _firstStackItemMS < uint.MaxValue - StackRefreshMS)
+                        case Utilities.MemoryPressure.Medium:
+                            trimCount = StackMediumTrimCount;
+                            break;
+                    }
+
+                    while (_count > 0 && trimCount-- > 0)
+                    {
+                        T[]? array = _arrays[--_count];
+                        Debug.Assert(array is not null, "No nulls should have been present in slots < _count.");
+                        _arrays[_count] = null;
+
+                        if (log.IsEnabled())
                         {
-                            // Give the remaining items a bit more time
-                            _firstStackItemMS += StackRefreshMS;
+                            log.BufferTrimmed(array.GetHashCode(), array.Length, id);
                         }
                     }
+
+                    _timestamp = _count > 0 ?
+                        _timestamp + StackRefreshMS : // Give the remaining items a bit more time
```

Comment:
Should this rather be `trimTicks / 4`? Otherwise, the `_timestamp` will be in future for `MemoryPressure.High` that will be treated as tickcount wrap around.